mesonbuild/interpreterbase/decorators.py

```diff
@@ -782,6 +782,64 @@ def log_usage_warning(self, tv: str, location: T.Optional['mparser.BaseNode']) -
         mlog.deprecation(*args, location=location)
 
 
+class FeatureExperimental(FeatureCheckBase):
+
+    """Checks for experimental features"""
+
+    feature_registry = {}
+
+    def __init__(self, feature_name: str, introduced_version: str, extra_message: str = '',
+                 stable_version: T.Optional[str] = None):
+        super().__init__(feature_name, introduced_version, extra_message)
+        self.stabilized = stable_version
+
+    @staticmethod
+    def get_warning_str_prefix(tv: str) -> str:
+        return 'Experimental features used:'
+
+    @staticmethod
+    def get_notice_str_prefix(tv: str) -> str:
+        return ''
+
+    @staticmethod
+    def check_version(target_version: str, feature_version: str) -> bool:
+        # For deprecation checks we need to return the inverse of FeatureNew checks
+        return not mesonlib.version_compare_condition_with_min(target_version, feature_version)
+
+    def log_usage_warning(self, tv: str, location: T.Optional['mparser.BaseNode']) -> None:
+        args = [
+            'Project targets', f"'{tv}'",
+            'and uses an experimental feature introduced in',
+            f"'{self.feature_version}':",
+            f'{self.feature_name}.',
+        ]
+        if self.stabilized:
+            args.append(f'which is stabalized in "{self.stabilized}"')
+        args.extend([
+            'Experimental features do not make forward or backward API guarantees,',
+            'and may be changed or removed in future releases, including point releases.'
+        ])
+        if self.extra_message:
+            args.append(self.extra_message)
+        mlog.warning(*args, location=location, fatal=False)
+
+    @classmethod
+    def single_use(cls, feature_name: str, introduced_version: str, subproject: SubProject,
+                   extra_message: str = '', location: T.Optional[mparser.BaseNode] = None,
+                   stable_version: T.Optional[str] = None) -> None:
+        cls(feature_name, introduced_version, extra_message, stable_version).use(subproject, location)
+
+    def use(self, subproject: 'SubProject', location: T.Optional['mparser.BaseNode'] = None) -> None:
+        tv = self.get_target_version(subproject)
+        # No target version
+        if not tv:
+            return
+        # I the feature is stabilized don't warn
+        if self.stabilized and mesonlib.version_compare(tv, f'>= {self.stabilized}'):
+            return
```

Comment:
`super().use()` already handles the `not tv` case, so we should not repeat its work here. The only purpose of checking this early is to avoid crashing on mesonlib.version_compare, which means that we should use:
```python
if tv and self.stabilized and mesonlib.version_compare(...)
```